### PR TITLE
Resolve path before initializing objwriter

### LINKED ISSUE #87961: ILC fails to initialize native object writer with path containing '../'
## Description

The ILC compiler encounters a failure when attempting to initialize the Native Object Writer with a path that contains `../`. When the path does not include a parent directory reference, the compiler initializes the writer correctly.

```
error : Unable to create file for /tmp/helix/working/ABCA092E/w/C3E60A32/e/apple_build/../publish/native/Program.o: No such file or directory [/private/tmp/helix/working/ABCA092E/w/C3E60A32/e/publish/ProxyProjectForAOTOnHelix.proj]
error : Fail to initialize Native Object Writer [/private/tmp/helix/working/ABCA092E/w/C3E60A32/e/publish/ProxyProjectForAOTOnHelix.proj]
  System.IO.IOException: Fail to initialize Native Object Writer
     at ILCompiler.DependencyAnalysis.ObjectWriter..ctor(String objectFilePath, NodeFactory factory, ObjectWritingOptions options) in /Users/miloskotlar/dotnet/runtime-ios/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:line 856
     at ILCompiler.DependencyAnalysis.ObjectWriter.EmitObject(String objectFilePath, IReadOnlyCollection`1 nodes, NodeFactory factory, ObjectWritingOptions options, IObjectDumper dumper, Logger logger) in /Users/miloskotlar/dotnet/runtime-ios/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:line 947
     at ILCompiler.Compilation.ILCompiler.ICompilation.Compile(String outputFile, ObjectDumper dumper) in /Users/miloskotlar/dotnet/runtime-ios/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs:line 523
     at ILCompiler.Program.Run() in /Users/miloskotlar/dotnet/runtime-ios/src/coreclr/tools/aot/ILCompiler/Program.cs:line 545
     at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass215_0.<.ctor>b__0(InvocationContext context) in /Users/miloskotlar/dotnet/runtime-ios/src/coreclr/tools/aot/ILCompiler/ILCo
```

### PR DESCRIPTION
**Pull request by am11**
Fix https://github.com/dotnet/runtime/issues/87961

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs [hunk 1]
```diff
@@ -850,10 +850,10 @@ public ObjectWriter(string objectFilePath, NodeFactory factory, ObjectWritingOpt
         {
             var triple = GetLLVMTripleFromTarget(factory.Target);
 
-            _nativeObjectWriter = InitObjWriter(objectFilePath, triple);
+            _nativeObjectWriter = InitObjWriter(Path.GetFullPath(objectFilePath), triple);
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:853]: What is the API that the object writer uses to open the file that makes it choke on `..`?

Comment (am11) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:853]: Seems like it ends up calling `open(2)` from https://github.com/dotnet/llvm-project/blob/8dad63a6a1ba42171420f2f6e24f2b5d0a5179ab/llvm/lib/Support/Unix/Path.inc#L1012, which should be able to handle the relative paths, so not 100% sure why it is failing for @kotlarmilos?

Locally on osx-arm64, it is working fine with main branch when I changed `-o:..` line in `artifacts/obj/coreclr/crossgen2/arm64/Debug/native/crossgen2.ilc.rsp` to a relative path `/Users/am11/projects/runtime/src/../artifacts/obj/coreclr/crossgen2/arm64/Debug/native/crossgen2.o`.

Comment (kotlarmilos) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:853]: I've encountered the failure on the CI when targeting iOS platforms in https://github.com/dotnet/runtime/pull/87773. The host platform is `osx-arm64`. I will try to reproduce it locally.

Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:853]: >  I will try to reproduce it locally

@kotlarmilos were you able to reproduce this?

Comment (kotlarmilos) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:853]: Sorry for the delayed reply. Yes, I can reproduce the issue locally on `osx-arm64`.

For example, in the sample app, the `NativeIntermediateOutputPath` is
```
/Users/miloskotlar/dotnet/runtime/src/mono/sample/iOS-NativeAOT/obj/native/
```
If I override `NativeIntermediateOutputPath` with
```
/Users/miloskotlar/dotnet/runtime/src/mono/sample/iOS-NativeAOT/obj/native/dummydir/../
```
the ILCompiler fails to initialize the Native Object Writer.

I made a mistake regarding the Helix host platform, it should be `osx-x64`. It appears that the issue is not specific to the architecture.

Comment (jkotas) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:853]: Does dummydir exist? If it does not exist, it is expected for this to fail.

Comment (kotlarmilos) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:853]: That's right, it doesn't exist. Probably that is the case with the Helix as well.

Comment (kotlarmilos) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:853]: @am11 If no further actions are expected, I suggest we close the issue and PR.

Thank you all for the assistance!

**Pull request closed without merging**